File: samples/Windows/java/TableManagerSamples/GetOffers/src/TableListener.java
package getoffers;

import com.fxcore2.*;

public class TableListener implements IO2GTableListener {
    private String mInstrument;
    private OfferCollection mOffers;

    // ctor
    public TableListener() {
        mOffers = new OfferCollection();
    }

    public void setInstrument(String sInstrument) {
        mInstrument = sInstrument;
    }

    // Implementation of IO2GTableListener interface public method onAdded
    public void onAdded(String sRowID, O2GRow rowData) {
    }

    // Implementation of IO2GTableListener interface public method onChanged
    public void onChanged(String sRowID, O2GRow rowData) {
        if (rowData.getTableType() == O2GTableType.OFFERS) {
            printOffer((O2GOfferTableRow)rowData, mInstrument);
        }
    }

    // Implementation of IO2GTableListener interface public method onDeleted
    public void onDeleted(String sRowID, O2GRow rowData) {
    }

    public void onStatusChanged(O2GTableStatus status) {
    }

    public void subscribeEvents(O2GTableManager manager) {
        O2GOffersTable offersTable = (O2GOffersTable)manager.getTable(O2GTableType.OFFERS);
        offersTable.subscribeUpdate(O2GTableUpdateType.UPDATE, this);
    }

    public void unsubscribeEvents(O2GTableManager manager) {
        O2GOffersTable offersTable = (O2GOffersTable)manager.getTable(O2GTableType.OFFERS);
        offersTable.unsubscribeUpdate(O2GTableUpdateType.UPDATE, this);
    }

    public void printOffers(O2GOffersTable offers, String sInstrument) {
        O2GOfferTableRow offerRow = null;
        O2GTableIterator iterator = new O2GTableIterator();
        offerRow = offers.getNextRow(iterator);
        while (offerRow != null) {
            printOffer(offerRow, sInstrument);
            offerRow = offers.getNextRow(iterator);
        }
    }

    public void printOffer(O2GOfferTableRow offerRow, String sInstrument) {
        Offer offer = mOffers.findOffer(offerRow.getOfferID());
        if (offer != null) {
            if (offerRow.isTimeValid() && offerRow.isBidValid() && offerRow.isAskValid()) {
                offer.setDate(offerRow.getTime());
                offer.setBid(offerRow.getBid());
                offer.setAsk(offerRow.getAsk());
            }
        } else {
            offer = new Offer(offerRow.getOfferID(), offerRow.getInstrument(),
                             offerRow.getDigits(), offerRow.getPointSize(),
                             offerRow.getTime(), offerRow.getBid(), offerRow.getAsk());
            mOffers.addOffer(offer);
        }
        if (sInstrument.isEmpty() || offerRow.getInstrument().equals(sInstrument)) {
            System.out.println(String.format("%s, %s, Bid=%s, Ask=%s", offer.getOfferID(), offer.getInstrument(), offer.getBid(), offer.getAsk()));
        }
    }
}

File: samples/Linux/netcore_for_Linux_x64_only/netcore1.0/TableManagerSamples/GetOffers/src/TableListener.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Text;
using System.Threading;
using fxcore2;

namespace GetOffers
{
    class TableListener : IO2GTableListener
    {
        private string mInstrument;
        private OfferCollection mOffers;

        /// <summary>
        /// ctor
        /// </summary>
        public TableListener()
        {
            mOffers = new OfferCollection();
        }

        public void SetInstrument(string sInstrument)
        {
            mInstrument = sInstrument;
        }

        #region IO2GTableListener Members

        // Implementation of IO2GTableListener interface public method onAdded
        public void onAdded(string sRowID, O2GRow rowData)
        {
        }

        // Implementation of IO2GTableListener interface public method onChanged
        public void onChanged(string sRowID, O2GRow rowData)
        {
            if (rowData.TableType == O2GTableType.Offers)
            {
                PrintOffer((O2GOfferTableRow)rowData, mInstrument);
            }
        }

        // Implementation of IO2GTableListener interface public method onDeleted
        public void onDeleted(string sRowID, O2GRow rowData)
        {
        }

        public void onStatusChanged(O2GTableStatus status)
        {
        }

        #endregion

        public void SubscribeEvents(O2GTableManager manager)
        {
            O2GOffersTable offersTable = (O2GOffersTable)manager.getTable(O2GTableType.Offers);
            offersTable.subscribeUpdate(O2GTableUpdateType.Update, this);
        }

        public void UnsubscribeEvents(O2GTableManager manager)
        {
            O2GOffersTable offersTable = (O2GOffersTable)manager.getTable(O2GTableType.Offers);
            offersTable.unsubscribeUpdate(O2GTableUpdateType.Update, this);
        }

        public void PrintOffers(O2GOffersTable offers, string sInstrument)
        {
            O2GOfferTableRow offerRow = null;
            O2GTableIterator iterator = new O2GTableIterator();
            while (offers.getNextRow(iterator, out offerRow))
            {
                PrintOffer(offerRow, sInstrument);
            }
        }

        public void PrintOffer(O2GOfferTableRow offerRow, string sInstrument)
        {
            Offer offer;
            if (mOffers.FindOffer(offerRow.OfferID, out offer))
            {
                if (offerRow.isTimeValid && offerRow.isBidValid && offerRow.isAskValid)
                {
                    offer.Date = offerRow.Time;
                    offer.Bid = offerRow.Bid;
                    offer.Ask = offerRow.Ask;
                }
            }
            else
            {
                offer = new Offer(offerRow.OfferID, offerRow.Instrument,
                                offerRow.Digits, offerRow.PointSize, offerRow.Time,
                                offerRow.Bid, offerRow.Ask);
                mOffers.AddOffer(offer);
            }
            if (string.IsNullOrEmpty(sInstrument) || offerRow.Instrument.Equals(sInstrument))
            {
                Console.WriteLine("{0}, {1}, Bid={2}, Ask={3}", offer.OfferID, offer.Instrument, offer.Bid, offer.Ask);
            }
        }
    }
}
